### PR TITLE
Add lident to C* for more accurate name handling; rework Genctypes

### PR DESCRIPTION
**Pull request by msprotz**
Hi Victor,

I would love for you to review my changes to Genctypes. As the pull request title says, I have been reworking C* to have long idents which eliminates a lot of (sometimes flawed) logic and makes it easy to perform tasks such as:
- recover the name of the file a given declaration will be emitted in
- match on the module name of a given declaration *without* having to resort to a reverse-map from C names to original names
- etc. etc.

This has, I believe, vastly simplified the handling of many idioms in this file. I also took the liberty of rewriting the graph-traversal logic to make it modular and clearly establish separate phases, data structures, and bits of logic. In particular, I believe I have completely separated the logic that determines which declarations should be bound via ctypes, from the actual OCaml syntax generation.

I also changed a bunch of your functions to clearly separate error handling -- now error handling is done at the top in the main traversal. If a declaration we ought to visit is unsupported for one reason or another, we just don't bother recursing into it.

I hope this separates the high-level logic from the nitty-gritty of OCaml syntax generation.

I have also changed the use of includes, replacing them with opens. I believe this avoids the diamond graph problem where there might be no simple way to avoid double definitions in scope.

Thanks,

Jonathan

### CODE REVIEW

File: src/GenCtypes.ml [hunk 1]
```diff
@@ -70,38 +70,47 @@ let mk_simple_app_decl (name: ident) (typ: ident option) (head: ident)
...
   | Some r -> r
   | None ->
-    if List.mem typ special_types then
+    if Hashtbl.mem special_types typ then
       default
-    else Warn.fatal_error "Type %s (in %s) not found in context and special handling not defined" typ location
+    else
+      Warn.fatal_error "Type %a (in %a) not found in context and special \
+        handling not defined"
+        plid typ plid location
```
Comment (victor-dumitrescu) [inline src/GenCtypes.ml]: This is now unused, can be deleted.

File: src/GenCtypes.ml [hunk 2]
```diff
@@ -213,308 +208,288 @@ and mk_struct_decl (k: structured) module_name name fields: structure_item list
...
- *   1) `compute_bundles` goes over each bundle and includes it in the list of bundles for which
- *      bindings will be generated if it contains any declarations originally from modules passed to `-ctypes`
- *   2) `compute_bindings` goes over each bundle and builds a table of declarations (`should_bind_decl`) according to
-       criteria (1) and (2) above
- *   3) `compute_dependencies` goes over each bundle and computes, for each declaration marked in the second pass,
- *     the set of types on which it depends. These types are also marked. Since the list of bundles (but not the modules
- *     within each bundle!) is topologically sorted, it is processed here in reverse order.
+  open_decls @ [m]
+
+(* We now need to visite the entire call-graph, and:
```
Comment (victor-dumitrescu) [inline src/GenCtypes.ml]: Typo (visit)

File: src/GenCtypes.ml [hunk 3]
```diff
@@ -213,308 +208,288 @@ and mk_struct_decl (k: structured) module_name name fields: structure_item list
...
+ * - generate OCaml bindings for: all the C* declarations whose lid matches a -ctypes
+ *   pattern, but also all C* types that are reachable through those declarations, and
+ * - compute accurate direct and transitive dependencies; the former for the sake of
+ *   generating include OCaml statements in the generate files, and the latter
```
Comment (victor-dumitrescu) [inline src/GenCtypes.ml]: Typo (generated files)

**Comment by msprotz** [comment]
Thanks for the review!

**Pull request merged**